prow/crier/controller_test.go

```diff
@@ -0,0 +1,231 @@
+	"k8s.io/test-infra/prow/client/clientset/versioned/fake"
+	listerv1 "k8s.io/test-infra/prow/client/listers/prowjobs/v1"
+	"k8s.io/test-infra/prow/kube"
+	"sync"
+	"testing"
+	"time"
+)
+
+const (
+	testTimeout    = 1 * time.Second
+	closeTimeLimit = 1 * time.Second
+)
+
+//Fake Reporter
+//Sets: Logic; Some series of test jobs should be reported, while others shouldn't be
+//Asserts: The correct jobs have Report called on them
+type triggerReporter struct {
+	reportedItem *bool
+}
+
+func (f triggerReporter) Report(pj *apiv1.ProwJob) ([]*apiv1.ProwJob, error) {
+	if f.reportedItem != nil {
+		*f.reportedItem = true
+	}
+	return nil, nil
+}
+
+func (f triggerReporter) GetName() string {
+	return "triggerReporter"
+}
+
+func (f triggerReporter) ShouldReport(pj *apiv1.ProwJob) bool {
+	return pj.Status.State == apiv1.TriggeredState
+}
+
+//Fake Informer
+//Sets: The Prow Job Test Cases
+//Asserts: The Controller registers for the queue to be populated
+type fakeInformer struct {
+	eventChannel chan string
+}
+
+func (f fakeInformer) Get(name string) (*apiv1.ProwJob, error) {
+	var pj *apiv1.ProwJob
+	switch name {
+	case "TestJob1":
+		pj = &apiv1.ProwJob{
+			Spec: apiv1.ProwJobSpec{
+				Job: "TestJob1",
+			},
+			Status: apiv1.ProwJobStatus{
+				State: apiv1.FailureState,
+			},
+		}
+	case "TestJob2":
+		pj = &apiv1.ProwJob{
+			Spec: apiv1.ProwJobSpec{
+				Job: "TestJob2",
+			},
+			Status: apiv1.ProwJobStatus{
+				State: apiv1.TriggeredState,
+			},
+		}
+	}
+
+	return pj, nil
+}
+
+func (f fakeInformer) ProwJobs(namespace string) listerv1.ProwJobNamespaceLister {
+	return f
+}
+
+func (f fakeInformer) Informer() cache.SharedIndexInformer {
+	return f
+}
+
+func (f fakeInformer) Lister() listerv1.ProwJobLister {
+	return f
+}
+
+func (f fakeInformer) AddEventHandler(handler cache.ResourceEventHandler) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Registered"
+	}
+}
+
+func (f fakeInformer) Run(stopCh <-chan struct{}) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Run"
+	}
+}
+
+func (f fakeInformer) AddEventHandlerWithResyncPeriod(handler cache.ResourceEventHandler, resyncPeriod time.Duration) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Registered"
+	}
+}
+
+func (fakeInformer) HasSynced() bool {
+	return true
+}
+
+func (fakeInformer) GetStore() cache.Store                                           { return nil }
+func (fakeInformer) GetController() cache.Controller                                 { return nil }
+func (fakeInformer) LastSyncResourceVersion() string                                 { return "" }
+func (fakeInformer) AddIndexers(indexers cache.Indexers) error                       { return nil }
+func (fakeInformer) GetIndexer() cache.Indexer                                       { return nil }
+func (fakeInformer) List(selector labels.Selector) (ret []*apiv1.ProwJob, err error) { return nil, nil }
+
+func TestController_processNextItem(t *testing.T) {
```

Comment:
Please prefer table driven testing

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/testing.md#unit-tests

Reply:
Thanks for this note; I think the table driven version is a stark improvement

Reply:
I think you're spending way to much time trying to test the behavior of c.queue.Get() which doesn't seem particularly interesting.

I would be inclined to refactor the controller to make this function trivial and pass the key to another method, which this unit tests.

Also I would be inclined to refactor usage like `c.informer.Lister().ProwJobs(namespace).Get(name)` into something more like `get(namespace, name) ProwJob` so that you are not faking a bunch of unnecessary methods